FHEM/10_SD_Rojaflex.pm

```diff
@@ -0,0 +1,639 @@
+			if ($attrValue !~ m/^[0-1]$/xms) { return "$name: Unallowed value $attrValue for the attribute bidirectional (must be 0 - 1)!" };
+		}
+		if ($attrName eq 'timeToClose' || $attrName eq 'timeToOpen') {
+			if ($attrValue !~ m/^\d{1,3}$/xms || $attrValue < 1) { return "$name: Unallowed value $attrValue for the attribute $attrName (must be 1 - 999)!" };
+		}
+	}
+	return;
+}
+
+sub Set {
+	my ($hash, $name, @a) = @_;
+	my $ioname = $hash->{IODev}{NAME};
+	my $na = scalar @a; # Anzahl in Array
+	my $cmd = $a[0];
+
+	return q(down:noArg stop:noArg up:noArg savefav:noArg gotofav:noArg clearfav:noArg pct:0,10,20,30,40,50,60,70,80,90,100) if ($cmd eq q(?));
+	return qq($name, no set command specified) if ($na < 1);
+	return qq($name, invalid set command) if (not exists $rev_codes{$cmd});
+	return qq($name, invalid parameter for command pct, must be 0-100) if ($cmd eq 'pct' && ($na < 2 || $a[1] !~ m/^\d+$/xms || $a[1] > 100));
+	return qq(Dummydevice $name: will not set data) if (IsDummy($name));
+
+	my $state;
+	my $motor = ReadingsVal($name, 'motor', 'stop');
+	my $cpos = ReadingsVal($name, 'cpos', undef);
+	my $tpos = ReadingsVal($name, 'tpos', 50);
+	if (!defined $cpos) {$cpos = $tpos};
+	if (AttrVal($name,'inversePosition',0) eq '1') {
+		$cpos = 100 - $cpos; # inverse position
+		$tpos = 100 - $tpos; # inverse position
+	}
+	my $timeToClose = AttrVal($name,'timeToClose',30);
+	my $timeToOpen = AttrVal($name,'timeToOpen',30);
+
+	if ($cmd eq 'pct') {
+		$tpos = $a[1];
+		if (AttrVal($name,'inversePosition',0) eq '1') {$tpos = 100 - $tpos}; # inverse position
+		if ($tpos eq '0') {$cmd = 'up'}; # Fahr hoch
+		if ($tpos eq '100') {$cmd = 'down'}; # Fahr runter
+		if ($tpos != $cpos) {
+			Log3 $name, 3, "$ioname: SD_Rojaflex set $name pct $tpos";
+			if ($tpos > 0 && $tpos < 100) {
+				my $duration;
+				if ($tpos > $cpos) { # Rolladen steht höher soll position
+					$cmd = 'down'; # Fahr runter
+					$duration = ($tpos - $cpos) * $timeToClose / 100;
+				}
+				if ($tpos < $cpos) { # Rolladen steht niedriger soll position
+					$cmd = 'up';# Fahr hoch
+					$duration = ($cpos - $tpos) * $timeToOpen / 100;
+				}
+				Log3 $name, 4, "$ioname: SD_Rojaflex set $name duration running time $duration s";
+				InternalTimer( (gettimeofday() + $duration), \&SD_Rojaflex_pctStop, $name );
+			}
+		} else {
+			$cmd = 'stop';
+		}
+	}
+
+	# Build msg and send it
+	my ($housecode,$channel) = split m/[_]/xms,$hash->{DEF};
+	my $msg = q(P109#08) . $housecode . sprintf('%X',$channel);
+	if ($cmd eq 'clearfav') {
+		$msg .= q(D); # gotofav
+	} else {
+		$msg .= $rev_codes{$cmd};
+	}
+	$msg .= q(A01);
+	if ($cmd eq 'clearfav') {
+		$msg .= q(D); # gotofav
+	} else {
+		$msg .= $rev_codes{$cmd};
+	}
+	$msg .= q(A);
+  my $sum = 0;
+  for (my $i = 7; $i < 20; $i += 2) {
+    $sum += hex(substr($msg, $i, 2));
+  }
+  $sum &= 0xFF;
+	$msg .= sprintf('%02X',$sum);
+	Log3 $name, 4, "$ioname: $name sendMsg=$msg";
+	for my $i (1 .. AttrVal($name, 'repetition', 1)) {
```

Comment:
Ich kenne diese Motoren nicht, aber es erscheint mir seltsam, dass egal welcher Set Befehl gesendet wird, immer eine Nachricht gesendet wird.

Die Anzahl an wiederholungen könnte man doch in $msg mit #$repetition mitgeben und muss das nicht über eine Schleife realisieren. 
Siehe SIGNALduino_Set_sendMsg

Reply:
Wenn ich dich richtig verstehe kann sendMsg die repetition als Attribut mitbekommen, dann ist das natürlich schön gekapselt.

Was meinst du mit deiner Frage zu den Motoren. Da sollte ich zu allem Rede und Antwort stehen können. Sie können halt bestimmte Befehle empfangen und diese muss man halt senden. Also wenn ich per Set (Runter) möchte, muss ich das halt schicken ...

Reply:
Fast. Einfach die Repetition an `$msg` `#R$repetitions` anhängen :)

Die Frage zu den Motoren bezog sich darauf, dass bei jedem Kommando etwas gesendet wird. 
Ich hatte halt nicht im Blick dass der Befehl in %refcodes steckt und war der Meinung es wird fast immer das gleiche gesendet.
Ist so wie es ist super gemacht :)

Reply:
> Fast. Einfach die Repetition an `$msg` `#R$repetitions` anhängen :)

Das funktioniert bei FSK (noch) nicht, da m.E. keine Möglichkeit besteht, eine Pause zwischen den Wiederholungen zu definieren.